File: bit.py
class Bit(object):
    def __init__(self, l):
        if isinstance(l, int):
            self.ar = [0 for _ in range(l+1)]
        else:
            self.ar = [0] + l

    def __getitem__(self, i):
        i += 1
        res = 0
        while i > 0:
            res += self.ar[i]
            i = i & (i - 1)
        return res

    def update(self, i, val):
        i += 1
        while i < len(self.ar):
            self.ar[i] += val
            i = i + (i & (-i))

File: cpp/bit.cpp
#include <stdint.h>
#include <vector>
#include <assert.h>

template<typename T>
class Bit {
public:
    Bit(int n): ar(n+1) {}
    T operator [](std::size_t i) {
        T res{};
        while (i > 0) {
            res += ar[i];
            i = i & (i - 1);
        }
        return res;
    }

    void update(int i, T val) {
        while (i < ar.size()) {
            ar[i] += val;
            i += i & (-i);
        }
    }
private:
    std::vector<T> ar;
};

int main() {
    using namespace std;
    int n = 1000;
    vector<int> vec(n);
    for (int i = 0; i < n; ++i) vec[i] = n - i;
    Bit<int> b(n);
    long long res = 0;
    for (int i = n-1; i >= 0; --i) {
        res += b[vec[i]];
        b.update(vec[i], 1);
    }
    assert(res == n * (n-1) / 2);
}
